examples/fluids/src/misc.c

```diff
@@ -465,3 +466,151 @@ PetscErrorCode IntArrayP2C(PetscInt num_entries, PetscInt **array_petsc, CeedInt
+  // -- Get Partition Statistics
+  PetscCall(PetscPrintf(comm, "  Partition:                             (min,max,median,max/median)\n"));
+  {
+    PetscInt *gather_buffer = NULL;
+    PetscInt  part_owned_dofs[3], part_local_dofs[3], part_ghost_dofs[3], part_neighbors[3];
+    PetscInt  median_index = comm_size % 2 ? comm_size / 2 : comm_size / 2 - 1;
+    if (!rank) PetscCall(PetscMalloc1(comm_size, &gather_buffer));
+
+    PetscCallMPI(MPI_Gather(&owned_dofs, 1, MPIU_INT, gather_buffer, 1, MPIU_INT, 0, comm));
+    if (!rank) {
+      PetscCall(PetscSortInt(comm_size, gather_buffer));
+      part_owned_dofs[0]             = gather_buffer[0];              // min
+      part_owned_dofs[1]             = gather_buffer[comm_size - 1];  // max
+      part_owned_dofs[2]             = gather_buffer[median_index];   // median
+      PetscReal part_owned_dof_ratio = (PetscReal)part_owned_dofs[1] / (PetscReal)part_owned_dofs[2];
+      PetscCall(PetscPrintf(
+          comm, "    Owned free %" PetscInt_FMT "-DoF nodes             : %" PetscInt_FMT ", %" PetscInt_FMT ", %" PetscInt_FMT ", %f\n", num_comp_q,
+          part_owned_dofs[0] / num_comp_q, part_owned_dofs[1] / num_comp_q, part_owned_dofs[2] / num_comp_q, part_owned_dof_ratio));
+    }
+
+    PetscCallMPI(MPI_Gather(&local_dofs, 1, MPIU_INT, gather_buffer, 1, MPIU_INT, 0, comm));
+    if (!rank) {
+      PetscCall(PetscSortInt(comm_size, gather_buffer));
+      part_local_dofs[0]             = gather_buffer[0];              // min
+      part_local_dofs[1]             = gather_buffer[comm_size - 1];  // max
+      part_local_dofs[2]             = gather_buffer[median_index];   // median
+      PetscReal part_local_dof_ratio = (PetscReal)part_local_dofs[1] / (PetscReal)part_local_dofs[2];
+      PetscCall(PetscPrintf(
+          comm, "    Local %" PetscInt_FMT "-DoF nodes                  : %" PetscInt_FMT ", %" PetscInt_FMT ", %" PetscInt_FMT ", %f\n", num_comp_q,
+          part_local_dofs[0] / num_comp_q, part_local_dofs[1] / num_comp_q, part_local_dofs[2] / num_comp_q, part_local_dof_ratio));
+    }
+
+    PetscInt num_remote_roots_total = 0, num_neighbor_ranks = 0;
+    {
+      PetscSF            sf;
+      PetscInt           nrranks, niranks;
+      const PetscInt    *roffset, *rmine, *rremote, *ioffset, *irootloc;
+      const PetscMPIInt *rranks, *iranks;
+      PetscCall(DMGetSectionSF(user->dm, &sf));
+      PetscCall(PetscSFGetRootRanks(sf, &nrranks, &rranks, &roffset, &rmine, &rremote));
+      PetscCall(PetscSFGetLeafRanks(sf, &niranks, &iranks, &ioffset, &irootloc));
+      for (PetscInt i = 0; i < nrranks; i++) {
+        if (rranks[i] == rank) continue;  // Ignore same-part global->local transfers
+        num_remote_roots_total += roffset[i + 1] - roffset[i];
+        num_neighbor_ranks++;
+      }
+      for (PetscInt i = 0; i < niranks; i++) {
+        if (iranks[i] == rank) continue;
+        num_neighbor_ranks++;
+      }
+    }
```

Comment:
I assume you were wanting all neighbors, right? Not just the number of neighbors owning ghost nodes?

Reply:
We don't want to double-count. We have neighbors that we receive from in the Bcast (global-to-local) and those we receive from in the Gather (local-to-global). Those numbers are different and it's possible for one to be tiny while the other is huge (but a good partitioner shouldn't leave you that way). It wouldn't hurt to print both, or if we want to save lines, just print the larger of the two.

Reply:
Ah, I guess I assumed that data would waterfall rather than be. I'll do both.

Reply:
Is it ok to assume that the current rank is included? Such that I can just use `nrranks`/`niranks` without the for loop?

Reply:
I think we want max and mean sizes so you'd have the loop anyway.

Reply:
I don't need a loop to get the per-rank statistics, just to get local value. The statistics are handled by the `MPI_Gather` and `PetscSortInt`.  The only reason I have a for loop right now is to exclude the remote roots on the current rank.

Reply:
We have several messages per rank so you must need the loop of you want per-message stats, not just per-rank.

Reply:
I'm not sure I follow. Assuming the commit I just pushed is correct, I don't see a reason why I couldn't do
```diff
-      for (PetscInt i = 0; i < niranks; i++) {
-        if (iranks[i] == rank) continue;
-        num_L2Gneighbor_ranks++;
-      }
+ num_L2Gneighbor_ranks = niranks - 1;
```

if `rank` is always present in `iranks`.

Reply:
That's just counting ranks, but I was thinking of message sizes (`roffset[i+1] - roffset[i]`).

Reply:
Ah, gotcha. Would we want to label them both as GlobalToLocal and LocalToGlobal directly then? Otherwise I can't think of an inverse name of "ghost".

Reply:
Naming things is always the hardest. What about "Owned messages" and "Ghost messages"? Or "Messages as root" and "Messages as leaf"?

Reply:
I feel like "root" and "leaf" are less intuitive unless you're really familiar with SF (and I'm not sure if I'd put myself in that camp).

How about "Owned boundary nodes" and "Ghost boundary nodes", relying on the notion of partition boundaries (which is really what we're trying to quantify/communicate here).

Reply:
s/boundary/interface/ is fine with me.